### PR TITLE
Extend the  ZKProofs module

### PR DESCRIPTION
**Pull request by drskalman**
Move the ZK Proofs sub-module to its own module. It adds more indepth material into ZK Proofs section so the participant get a feeling about the magic going behind generating and verifying zk-proofs and taste of how to make one perhaps for their application in future.

This currently kept in a different folder but the plan is to move it to common cryptography folder when it is ready to be merged.

### CODE REVIEW

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+
+(we are working on substrate support for these and will use them for protocols)
+
+---
+
+## What is a ZK Proof?
+
+- A prover wants to convince a verifier that something is true without revealing why it is true.
+
+- They are mostly interactive protocols, but mostly we'll turn them  non-interactive and deal with their non-interactive variety.
```
Comment (naterarmstrong) [inline syllabus/1-Cryptography/zk-proofs/slides.md:40]: Students don't know the definitions of interactive/non-interactive at this point, so probably add a speaker note to explain the distinction briefly before you get to the proper definition later on.

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+
+---
+
+## zk-SNARK
+
+**Z**ero-**K**nowledge **S**uccinct **N**on-interactive **Ar**gument of **K**nowledge
+
+- **Zero knowledge** - the proof reveals nothing about the witness that was not revealed by the statement itself.
+- **Succinct** - the proof is small
+- **Non-interactie** - Does not  require live interaction between prover and the verifier.
```
Comment (naterarmstrong) [inline syllabus/1-Cryptography/zk-proofs/slides.md:114]:
```suggestion
- **Non-interactive** - Does not require live interaction between the prover and verifier.
```

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+Notes:
+
+ - We are going to talk about the elementary math behind zk-SNARKs.
+ - The goal is to familiarize you with the magic behind the zk-SNARKs.
+ - Don't worry if you are not able to follow. Not essential.
+---
+
+## Making a SNARK out of the factorization problem
+- The trick  is to transform our problem of proving the knowledge of factors (witnesses) into:
+- A problem of knowledge of a certain polynomials.
```
Comment (naterarmstrong) [inline syllabus/1-Cryptography/zk-proofs/slides.md:203]:
```suggestion
- A problem of knowledge of certain polynomials.
```

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+- Then verifier could ask me questions about those polynomials, and if the prover answers correctly,
+- The verifier could be fairly confidant that the prover knows that polynomial hence also the witness.
```
Comment (naterarmstrong) [inline syllabus/1-Cryptography/zk-proofs/slides.md:205]:
```suggestion
- The verifier could be fairly confident that the prover knows that polynomial hence also the witness.
```

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+
+---
+
+## Making an SNARK for knowledge of factors problem
+- A routine way of to turning the problem into a polynomial is:
+1. To represents our problem into an arithmetic circuit.
+2. Then are algorithms such as PLONK for representing the circuit as few univariate polynomials.
```
Comment (naterarmstrong) [inline syllabus/1-Cryptography/zk-proofs/slides.md:212]: There are? Then use? Not sure what this was supposed to be

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+---
+
+## Overflow prevention
+- Our polynomials are all defined modulo some prime $p$
+- We need to prevent the prover from fooling us with a factorization like 
+
+    $r \times s = N + q\times p$ where $q \neq 0$.
+
+- This happens if $r$ and $s$ are two big and $r \times s$ overflows over $p$
```
Comment (naterarmstrong) [inline syllabus/1-Cryptography/zk-proofs/slides.md:234]:
```suggestion
- This happens if $r$ and $s$ are too big and $r \times s$ overflows over $p$
```

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+- We should make sure that $r$ and $s$ are small.
+- We use binary decomposition for that.
+- $r = r_{0} + 2r_{1} + 4r_{2}$ where 
+- We should only allow them use 0 or1 for $r_{i}$'s so they need to satisfies:
```
Comment (naterarmstrong) [inline syllabus/1-Cryptography/zk-proofs/slides.md:238]:
```suggestion
- We should only allow them to use 0 or 1 for $r_{i}$'s so they need to satisfy:
```

**Review by naterarmstrong — APPROVED** [review]
No major feedback, overall looks like a pretty cool (albeit math-heavy) lecture on zk proofs. The demos are great. I left some feedback primarily in the form of typos.

Make sure to remember to remove the ZK section of exotic primitives, as this now has its own lecture.

File: syllabus/1-Cryptography/zk-proofs/interpolation.sage [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+q_m_values = [ 1,  0,  0,  1,  1,  1,  1,  0,  0,  1,  1,  1,  1 ]
+q_c_values = [-n,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0, -1 ]
+
+r = 7
+s = 3
+
+
+def compute_input_output_poly_values(secret_factor):
+   #import pdb
+   #pdb.set_trace()
```
Comment (coax1d) [inline syllabus/1-Cryptography/zk-proofs/interpolation.sage:18]: Are these two lines still necessary?

File: syllabus/1-Cryptography/zk-proofs/slides.md (hunk 1 above)
Comment (coax1d) [inline syllabus/1-Cryptography/zk-proofs/slides.md:40]:
```suggestion
- They are mostly interactive protocols, but mostly we'll turn them non-interactive and deal with their non-interactive variety. 
```

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+---
+
+## Examples
+---
+
+## ZK Proof Example
+
+_Example:_ Schnorr signatures are ZK Proofs
+- $(s, e)$ such that $s = (\textrm{Random blinding factor }r) - \\ \textrm{Private Key} \times e$
+- They show that the prover knows the private key (the discrete log of the public key) without revealing anything about it.
```
Comment (coax1d) [inline syllabus/1-Cryptography/zk-proofs/slides.md:72]: I think you should tread lightly on this. Even mentioning discrete log I dont even think is necessary. I think it can just be generalized to the prover contains knowledge of a secret key and can prove it to the verifier without revealing it.

Something to this affect but if you do decide to continue just remember to tread lightly.

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+
+- Prover wants to prove that they know $N = r \times s$  without revealing $r$ or $s$.
+
+- The Prover should convince us they know the two non-trivial integer $r$ and $s$ that is $r,s \neq 1$ such that:
+
+ $r\times s = N$ 
+
+- Without revealing any other information about $r$ or $s$.
+
+- The statement is $N$ and the witnesses are $r$ and $s$.
```
Comment (coax1d) [inline syllabus/1-Cryptography/zk-proofs/slides.md:89]: In this example I think the math is appropriate because you are giving a concrete example using factorization.

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+
+  - ... Without revealing anything about the witness (_ZK_)
+
+- With a small proof even if the witness is large (_succinctness_)
+
+---
+
+## What can we show?
+
+- There are many SNARK schemes to produce succinct ZK proofs of knowledge (_ZK-SNARKs_) for every NP relation. We concentrate on PLONK in this course.
```
Comment (coax1d) [inline syllabus/1-Cryptography/zk-proofs/slides.md:133]: Just be ready to explain NP in a succinct way :)

No pun intended there.

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+- The verifier is a blockchain: very expensive data and computation costs.
+
+- Layer 2s using ZK rollups
+
+Notes:
+
+Of which Ethereum has many, ZKsync, ZKEVM etc.
+Polkadot already scales better!
+
+---
```
Comment (coax1d) [inline syllabus/1-Cryptography/zk-proofs/slides.md:171]: One of the original requirements of this module was to have no mention of blockchains. Just an FYI. We are building up to the blockchain datastructure. Going into Layer 2's is too premature.

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 13]
```diff
@@ -0,0 +1,446 @@
...
+
+## Privacy
+
+<pba-flex center>
+
+A user has private data, but we can show<br />publicly that this private data is correctly used.<br />
+An example would a private cryptocurrency:
+
+- Keep who pays who secret
+- Keep amounts secret, <br /> _But show they are positive!_
```
Comment (coax1d) [inline syllabus/1-Cryptography/zk-proofs/slides.md:181]: This I think is ok though. It is basic enough at this point. We do mention Bitcoin in some of the early hashing lectures as well. I remember Gav doing so aswell.

File: syllabus/1-Cryptography/zk-proofs/slides.md [hunk 14]
```diff
@@ -0,0 +1,446 @@
...
+ 
+ $(r-1)*\frac{1}{r - 1} = 1 \Rightarrow r\frac{1}{r - 1} - \frac{1}{r - 1} = 1$
+ 
+ $Q_l\times a + Q_r\times b + Q_o\times c + Q_m\times a\times b + Q_c = 0$ 
+
+<img style="height: 500px; padding-left:100px" src="./img/gate-table-left-input-less-than-8-and-not-1.png" />
+
+---
+
+## Gate table for right input to be integer and not 1
```
Comment (coax1d) [inline syllabus/1-Cryptography/zk-proofs/slides.md:326]:
```suggestion
## Gate table for the right input to be an integer and not 1
```

**Review by coax1d — COMMENTED** [review]
The basic definitions of ZK I think are done well to give some good understanding. But there are some things I would caution and emphasis.

1. Remember that basic math concepts or definitions will 100% be non-obvious to a majority of attendees. For example the idea of `NP` or even Factorization(I know this sounds silly but remember they are not math people in the slightest)

What are the repercussions of this?
If you simply skip over these things during the lecture people will be left scratching their head as to what that mathy thing meant thus distracting them from staying on task.

My suggestion is if you are going to mention `Discrete Log` or `NP` etc. That you quickly address the base definition succinctly.. or you can simply tell them "Hey look I am just mentioning this because it is mathematically correct but dont worry if you dont understand the full definition it isnt a big deal."

2. With any actual math just remember that they might not know even the simplest of things I would honestly almost assume math illiteracy. If you are fully understanding of that and still think you can manage then I will trust you to do things correctly.

But just remember what is the consequence of just ignoring the previous statement?
Basically you lose people and they tune out and it isn't a great experience. I think again all of this is doable but you have to be very in tune with where they are at pause frequently for questions and somehow not get derailed from finishing the lecture.

Overall though I like the example and walk through. I will be interested to see Gav's feedback on this.

File: syllabus/1-Cryptography/zk-proofs/slides.md (hunk 11 above)
Comment (drskalman) [reply syllabus/1-Cryptography/zk-proofs/slides.md:138]: I thnk all mentions of NP are inherited from @AlistairStewart 's slide. I could remove them if there is a consensus on that.

File: syllabus/1-Cryptography/zk-proofs/slides.md (hunk 12 above)
Comment (drskalman) [reply syllabus/1-Cryptography/zk-proofs/slides.md:176]: I do not believe I have added any mention of blockchain, they are all from previous slides.

**Comment by drskalman** [comment]
> I would honestly almost assume math illiteracy.

We learned about factors when we were about 9 or 10. Majority of the math is polynomial arithmetics which we learned at age 13. I'd think that assuming a year or two of post-primary education is safe for the majority of the participants. 

There are few mentions of module arithmetics, which is still high school level. And there is binary rep which shouldn't be strange for a coder. 

The most sophisticated tool is Lagrange interpolation which I'm treating as a blackbox. The same goes with KZG commitment. 

I removed DL. We could remove NP.

File: syllabus/1-Cryptography/zk-proofs/slides.md (hunk 11 above)
Comment (coax1d) [reply syllabus/1-Cryptography/zk-proofs/slides.md:138]: I think it is fine to keep in. I am just leaving a note that you should not assume everyone is familiar is all.

File: syllabus/1-Cryptography/zk-proofs/slides.md (hunk 12 above)
Comment (coax1d) [reply syllabus/1-Cryptography/zk-proofs/slides.md:176]: Up to you. I am just giving you what was told to us.

**Comment by coax1d** [comment]
I agreed with the factorization being appropriate based on my previous comments. My overall comment was just based on my experience from past PBAs. I like the level of math I just was wanting you to be aware is all. If you are aware than I think you have plenty of experience of how to proceed correctly and have my full support. Looking forward to teaching the module with you.

**Comment by AlistairStewart** [comment]
On NP, the students don't need to know what it is if they don't already. The key thing is to explain the API, that verifying is a procedure that takes the input and witness and returns true or false. (In practice we don't care if it is polynomial time, it needs to be concretely efficient. )

**Comment by drskalman** [comment]
The check fails for not finding `merkle-tree-activity` but it is there so I don't know why but I'm going to merge nonetheless.
